### PR TITLE
NEW: adds checksums to archive structure

### PR DESCRIPTION
**Pull request by ebolyen**
This commit adds a 'checksums.md5' file to the root of an archive.
The file is formatted with the md5sum format:

\<hex-hash>\<space>\<space or asterisk>\<escaped filename>\<newline>

Filename escaping handles backslash and newline. In the event those are
present, the backslash is doubled and the newline becomes `\n`.
Additionally if escaping is required, then an additional backslash
precedes the \<hex-hash> to indicate the filename is escaped.
See:
https://www.gnu.org/software/coreutils/manual/html_node/md5sum-invocation.html
for more details.

**TODO**:

- [x] Add tests

### CODE REVIEW

**Comment by ebolyen** [comment]
Ok this should be ready for review.

**Comment by ebolyen** [comment]
image

**Comment by ebolyen** [comment]
@gregcaporaso @thermokarst @nbokulich,

Question:

Should we copy the `checksums.md5` file to `/provenance/`? That would have the effect of making this file inherited by downstream artifacts (the checksum was already performed, so we would have the fingerprint of the original `/data/` directory for basically free). 

The downside is, the checksums will refer to an archive that no longer exists (from the perspective of the child), and the filepaths will be wrong/unusable with `md5sum`. We also would require some kind of tooling to really make any use of that data, but it would be interesting.

Another option would be to save that for a rainy day (perhaps when we have some notion of "who" did a thing with optional digital signatures? Then a mechanism like this would be almost required...)

**Comment by gregcaporaso** [comment]
> Should we copy the checksums.md5 file to /provenance/? 

@ebolyen, for the reasons you noted I think you should leave that one for a rainy day.

File: qiime2/core/archive/tests/test_archiver.py [hunk 1]
```diff
@@ -331,6 +337,59 @@ def test_parse_uuid_invalid(self):
...
+        # version (which is an unknown commit sha1), so just check name
+        self.assertEqual(list(diff.removed.keys()), ['VERSION'])
+        self.assertEqual(diff.changed,
+                         {'data/ints.txt': ('c0710d6b4f15dfa88f600b0e6b624077',
+                                            'f47bc36040d5c7db08e4b3a457dcfbb2')
+                          })
+
+    def setup_old_archiver(self):
+        class OldArchiver(Archiver):
+            # Quickest way to get the old code running
```
Comment (thermokarst) [inline qiime2/core/archive/tests/test_archiver.py]: Actually...

https://github.com/qiime2/qiime2/pull/198

File: qiime2/core/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2018, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+
+class ValidationError(Exception):
```
Comment (thermokarst) [inline qiime2/core/exceptions.py]: https://youtu.be/NGrLb6W5YOM

File: qiime2/core/archive/tests/test_archiver.py [hunk 3]
```diff
@@ -331,6 +338,44 @@ def test_parse_uuid_invalid(self):
...
+        self.assertEqual(diff.removed, {})
+        self.assertEqual(diff.changed, {})
+
+    def test_checksums_mismatch(self):
+        with (self.archiver.root_dir / 'data' / 'ints.txt').open('w') as fh:
+            fh.write('999\n')
+        with (self.archiver.root_dir / 'tamper.txt').open('w') as fh:
+            fh.write('extra file')
+
+        (self.archiver.root_dir / 'VERSION').unlink()
```
Comment (thermokarst) [inline qiime2/core/archive/tests/test_archiver.py]: spooky 👻

File: qiime2/sdk/tests/test_result.py [hunk 4]
```diff
@@ -218,6 +221,37 @@ def test_artifact_has_metadata_false(self):
...
+
+        artifact.validate()
+        self.assertTrue(True)  # Checkpoint
+
+    def test_validate_artifact_bad(self):
+        artifact = Artifact.import_data('IntSequence1', [1, 2, 3, 4])
+        with (artifact._archiver.root_dir / 'extra.file').open('w') as fh:
+            fh.write('uh oh')
+
+        with self.assertRaisesRegex(exceptions.ValidationError, 'extra\.file'):
```
Comment (thermokarst) [inline qiime2/sdk/tests/test_result.py]:
```suggestion
        with self.assertRaisesRegex(exceptions.ValidationError, r'extra\.file'):
```

File: qiime2/sdk/tests/test_result.py [hunk 5]
```diff
@@ -218,6 +221,37 @@ def test_artifact_has_metadata_false(self):
...
+        self.assertTrue(True)  # Checkpoint
+
+    def test_validate_vizualization_bad(self):
+        visualization = Visualization._from_data_dir(
+             self.data_dir, self.make_provenance_capture())
+
+        with (visualization._archiver.root_dir / 'extra.file').open('w') as fh:
+            fh.write('uh oh')
+
+        with self.assertRaisesRegex(exceptions.ValidationError, 'extra\.file'):
```
Comment (thermokarst) [inline qiime2/sdk/tests/test_result.py]:
```suggestion
        with self.assertRaisesRegex(exceptions.ValidationError, r'extra\.file'):
```

File: qiime2/sdk/tests/test_result.py [hunk 6]
```diff
@@ -218,6 +221,37 @@ def test_artifact_has_metadata_false(self):
...
+
+        artifact.validate()
+        self.assertTrue(True)  # Checkpoint
+
+    def test_validate_artifact_bad(self):
+        artifact = Artifact.import_data('IntSequence1', [1, 2, 3, 4])
+        with (artifact._archiver.root_dir / 'extra.file').open('w') as fh:
+            fh.write('uh oh')
+
+        with self.assertRaisesRegex(exceptions.ValidationError, r'extra\.file'):
```
Comment (thermokarst) [inline qiime2/sdk/tests/test_result.py]:
```suggestion
        with self.assertRaisesRegex(exceptions.ValidationError,
                                    r'extra\.file'):
```

File: qiime2/sdk/tests/test_result.py [hunk 7]
```diff
@@ -218,6 +221,37 @@ def test_artifact_has_metadata_false(self):
...
+        self.assertTrue(True)  # Checkpoint
+
+    def test_validate_vizualization_bad(self):
+        visualization = Visualization._from_data_dir(
+             self.data_dir, self.make_provenance_capture())
+
+        with (visualization._archiver.root_dir / 'extra.file').open('w') as fh:
+            fh.write('uh oh')
+
+        with self.assertRaisesRegex(exceptions.ValidationError, r'extra\.file'):
```
Comment (thermokarst) [inline qiime2/sdk/tests/test_result.py]:
```suggestion
        with self.assertRaisesRegex(exceptions.ValidationError,
                                    r'extra\.file'):
```

**Pull request merged**